lib/hash.h

```diff
@@ -2,42 +2,282 @@
+template <typename T>
+struct Hasher<T *> {
+    // FIXME: better use std::bit_cast from C++20
+    size_t operator()(T *val) const { return Hash()(reinterpret_cast<std::uintptr_t>(val)); }
+};
+
 template <typename T>
-auto murmur(const T &obj, typename std::enable_if<std::is_standard_layout<T>::value &&
-                                                  !std::is_pointer<T>::value>::type * = nullptr)
-    -> decltype(murmur(reinterpret_cast<const void *>(&obj), sizeof(T))) {
-    return murmur(reinterpret_cast<const void *>(&obj), sizeof(T));
+struct Hasher<std::unique_ptr<T>> {
+    size_t operator()(const std::unique_ptr<T> &val) const { return Hash()(val.get()); }
+};
+
+template <typename T>
+struct Hasher<std::shared_ptr<T>> {
+    size_t operator()(const std::shared_ptr<T> &val) const { return Hash()(val.get()); }
+};
+
+template <class Iter, class Hash = std::hash<typename std::iterator_traits<Iter>::value_type>>
+uint64_t hash_range(Iter begin, Iter end, uint64_t hash = 0, Hash Hasher = Hash()) {
+    for (; begin != end; ++begin) hash = hash_combine(hash, Hasher(*begin));
+
+    return hash;
+}
+
+template <class Hasher>
+inline size_t hash_combine_generic(const Hasher &) noexcept {
+    return 0;
 }
-}  // namespace Hash
+
+template <class Hasher, typename T, typename... Types>
+size_t hash_combine_generic(const Hasher &h, const T &t, const Types &...ts) {
+    size_t seed = h(t);
+    if (sizeof...(ts) == 0) return seed;
+
+    size_t remainder = hash_combine_generic(h, ts...);
+    if constexpr (sizeof(size_t) == sizeof(uint32_t))
+        return static_cast<size_t>(
+            hash_combine(static_cast<uint64_t>(seed) << 32, static_cast<uint64_t>(remainder)));
+    else
+        return static_cast<size_t>(hash_combine(seed, remainder));
+}
+
+template <typename T, typename... Types>
+size_t hash_combine(const T &t, const Types &...ts) {
+    return hash_combine_generic(Detail::StdHasher{}, t, ts...);
+}
+
+namespace Detail {
+template <size_t index, typename... Types>
+struct TupleHasher {
+    size_t operator()(std::tuple<Types...> const &val) const {
+        return hash_combine(TupleHasher<index - 1, Types...>()(val), std::get<index>(val));
+    }
+};
+
+template <typename... Types>
+struct TupleHasher<0, Types...> {
+    size_t operator()(std::tuple<Types...> const &val) const {
+        return hash_combine(std::get<0>(val));
+    }
+};
+}  // namespace Detail
 }  // namespace Util
 
+namespace std {
+template <typename T1, typename T2>
+struct hash<std::pair<T1, T2>> {
+    size_t operator()(const std::pair<T1, T2> &x) const {
+        return Util::hash_combine(x.first, x.second);
+    }
+};
+
+template <typename... Types>
+struct hash<std::tuple<Types...>> {
+ public:
+    size_t operator()(std::tuple<Types...> const &key) const {
+        Util::Detail::TupleHasher<sizeof...(Types) - 1, Types...> hasher;
+
+        return hasher(key);
+    }
+};
+}  // namespace std
```

Comment:
I'm a bit worried about this. Isn't it forbidden by the standard to specialize standard classes for other standard (and builtin) types (i.e. it is all fine with specializing `std::hash<>` for our `cstring`, but we should not specialize it for anything that the std defines)?

I would suggest we are just explicit and use the `Util::Hash` as hasher when we need to hash tuples. Possible problem is that in that case we could be that we would not be able to hash nested things that have `std::hash` but not `Util::Hash`... maybe `Util::Hash` can fallback to `std::hash`? The template specialization rules _might_ be actually doing the right think if you define the non-specialized struct to defer to `std::hash`.

Reply:
> I'm a bit worried about this. Isn't it forbidden by the standard to specialize standard classes for other standard (and builtin) types (i.e. it is all fine with specializing std::hash<> for our cstring, but we should not specialize it for anything that the std defines)?

That's exactly one of the points of aforementioned proposals. It was proposed in N3983 that the standard library would provide such specializations, but ... nothing happened. One of the main reasons is that there is no standard way to combine hashes and therefore everything stalled and became grey zone. I can certainly remove those specializations, it's not a big deal. Overall, specializations in `std` namespace are considered as mistakes in standard as rules what is permitted and what is not are confusing at least, but it's too late to fix it :) Maybe in C++42.

I'd rather not fallback `Util::Hash` to `std::hash`. Otherwise we might lose possibility to define latter in terms of former due to cyclic dependencies.

Reply:
Note that removing specializations for pairs and tuples will make impossible `hash_combine` as well:
```c++
    auto a = std::make_pair(1, 2);
    auto b = std::make_pair(3, 4);
    auto c = std::make_pair(1, 2);
    auto d = std::make_pair(2, 1);
    EXPECT_EQ(Util::hash_combine(a), Util::hash_combine(c));
    EXPECT_NE(Util::hash_combine(b), Util::hash_combine(c));
    EXPECT_NE(Util::hash_combine(d), Util::hash_combine(c));

    EXPECT_EQ(Util::hash_combine(a, b), Util::hash_combine(c, b));
    EXPECT_NE(Util::hash_combine(a, b), Util::hash_combine(b, a));
```
will stop working. One would need either use `hash_combine_generic` or disassemble pair / tuple.

Reply:
OK, I guess we can remove them when the standard finally defines them. The ambiguity should be quite clear at that point so hopefully it will not be a big mess.

Reply:
Yes, we can just get rid of them, it's easy :)